### DISCUSSION TITLE
CPU Temperature data not showing in Powershell via DLL

### DISCUSSION THREAD

**Discussion by iotaukm** [LibreHardwareMonitor/LibreHardwareMonitor#1358] [General]
Im trying to get various metrics of a windows system using LibreHardwareMonitor, and im limited to cli. Since i can load the LibreHardwareMonitorLib.dll file (as seen [here](https://stackoverflow.com/questions/76867278/ssd-temps-via-libre-hardware-monitor)), i decided to go this route. See below a modified version of the snippet:

```
cls
$dll = "LibreHardwareMonitorLib.dll"

Unblock-File -LiteralPath $dll
Add-Type -LiteralPath $dll
$monitor = [LibreHardwareMonitor.Hardware.Computer]::new()
$monitor.IsCPUEnabled = $true
$monitor.IsStorageEnabled = $true

$monitor.Open()
foreach ($sensor in $monitor.Hardware.Sensors) {
    write-host $sensor.Name
    write-host $sensor.SensorType
    write-host $sensor.Value
}
$monitor.Close()
```

I am using the same script, but am not getting all metrics. I can get cpu loads, but not clock speed or temperatures (which is my primary requirement).

See output below:

```
CPU Core #1 Thread #1
Load
99.38645
CPU Core #1 Thread #2
Load
23.83488
CPU Core #2 Thread #1
Load
31.51664
CPU Core #2 Thread #2
Load
69.42992
CPU Core #3 Thread #1
Load
0
CPU Core #3 Thread #2
Load
0
CPU Core #4 Thread #1
Load
0
CPU Core #4 Thread #2
Load
0
CPU Total
Load
10.33803
CPU Core Max
Load
99.38645
CPU Core #1
Temperature

CPU Core #2
Temperature

CPU Core #3
Temperature

CPU Core #4
Temperature

CPU Package
Temperature

CPU Core #1 Distance to TjMax
Temperature

CPU Core #2 Distance to TjMax
Temperature

CPU Core #3 Distance to TjMax
Temperature

CPU Core #4 Distance to TjMax
Temperature

Core Max
Temperature

Core Average
Temperature

CPU Core #1
Clock

CPU Core #2
Clock

CPU Core #3
Clock

CPU Core #4
Clock

```

**Comment by iotaukm** [comment]
Update:

I tried to build a console app in visual studio, and i followed the sample code by the developer. Im getting the same issue. Temperature stats do not show. Again, keep in mind that the GUI shows temperatures. But i cant get them any where else.

Even weirder.......i ran and re-ran the code.....multiple times. Build and re-build the application. Somehow.......ive ended up with one build where the temperature shows. Ive saved this DLL separately, but i cannot use this as it just dumps all the system data. I need directed outputs.

After i just re-built the solution, the temps vanished.

I have no clue whats going on at this point

**Comment by RodPaDev** [comment]
Possibly related to this in my case is that it isn's showing the CCD1 sensor even though the GUI shows it. 
It worked fine yesterday and nothing has changed since. I'm using the 0.9.3 version provided via Nuget (same as GUI is using).

```csharp

using LibreHardwareMonitor.Hardware;


internal class Program {
    private static void Main(string[] args) {
        var computer = new Computer {
            IsCpuEnabled = true,
        };
        computer.Open();
        foreach(var h in computer.Hardware) {
            foreach (var s in h.Sensors) {
                Console.WriteLine(s.Identifier);
            }
        }
    }
}
```

**Comment by RodPaDev** [comment]
@iotaukm Not sure if applicable to you but I tried running Visual Studio as Administrator and it started working. So you may need to run your executable in Administrator. Hopefully this helps you!

**Comment by QuackSilvr** [comment]
Hi,
after getting the same result as you initially, I followed the Code from the Readme to alter your code and it works.
Please know, that this is my first PS script and I don't know what I'm doing, I just tried to translate the code from the Readme into PS script (using microsoft help pages)

```
cls
$dll = "LibreHardwareMonitorLib.dll"

Unblock-File -LiteralPath $dll
Add-Type -LiteralPath $dll


class UpdateVisitor : LibreHardwareMonitor.Hardware.IVisitor
{

    [void] VisitComputer([LibreHardwareMonitor.Hardware.IComputer]$computer)
    {
        $computer.Traverse($this);
    }
    [void] VisitHardware([LibreHardwareMonitor.Hardware.IHardware]$hardware)
    {
        $hardware.Update();
        foreach ($subHardware in $hardware.SubHardware){
			$subHardware.Accept($this);
		}
    }
    [void] VisitSensor([LibreHardwareMonitor.Hardware.ISensor]$sensor) { }
    [void] VisitParameter([LibreHardwareMonitor.Hardware.IParameter]$parameter) { 	}
}



$monitor = [LibreHardwareMonitor.Hardware.Computer]::new()

$monitor.IsCPUEnabled = $true
$monitor.IsStorageEnabled = $true


$monitor.Open()

$monitor.Accept([UpdateVisitor]::new());

foreach ($hardware in $monitor.Hardware) {
	foreach ($sensor in $hardware.Sensors) {
		write-host $sensor.Name
		write-host $sensor.SensorType
		write-host $sensor.Value
	}
}
$monitor.Close()
```
